services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/client/S3ClientComposer.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.client;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.client.config.SdkClientConfiguration;
+import software.amazon.awssdk.core.client.config.SdkClientOption;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.S3Configuration;
+import software.amazon.awssdk.services.s3.internal.crossregion.S3CrossRegionAsyncClient;
+import software.amazon.awssdk.services.s3.internal.crossregion.S3CrossRegionSyncClient;
+import software.amazon.awssdk.utils.ConditionalDecorator;
+
+@SdkInternalApi
+public class S3ClientComposer {
```

Comment:
Maybe we need a class hierarchy here? Instead of using static method calls, we could have

```
interface ClientComposer<T extends AwsClient> {
    T composeSync(T base, SdkClientConfiguration clientConfiguration);
    T composeAsync(T base, SdkClientConfiguration clientConfiguration);
}
```
And then have `S3ClientComposer` implement that interface, and then have the codegen create a new instance of the correct implementation, based on the `model.clientComposerClassName()`.

I say that because there is no way of making sure that in the codegen when we call

```
if (model.clientComposerClassName().isPresent()) {
        builder.addStatement("return $T.composeAsync(client, clientConfiguration)",
                         PoetUtils.classNameFromFqcn(model.clientComposerClassName().get()));
```

if we add other "Composer", there is no way currently to enforce to write the correct method name which is called in the codegen.

Reply:
I briefly explore that route but didn't carry it through, thanks for pointing it out. Adding an interface.